src/main/java/teammates/common/util/retry/Retryable.java

```diff
@@ -1,14 +1,26 @@
-package teammates.test.driver.retry;
+package teammates.common.util.retry;
 
 /**
  * Represents a task that can be retried.
  * @param <T> Result type.
- * @param <E> Throwable type.
+ * @param <E> Throwable type for signalling that the task should not be retried.
  */
 public abstract class Retryable<T, E extends Throwable> {
 
     protected String name;
 
+    /**
+     * An optional final message to show in the {@link MaximumRetriesExceededException} thrown
+     * should the task fail after maximum retries.
+     */
+    protected String finalMessage;
+
+    /**
+     * An optional final object to embed in the {@link MaximumRetriesExceededException} thrown
+     * should the task fail after maximum retries.
+     */
+    protected Object finalData;
```

Comment:
The user might want to embed other types of data in the exception than the return type. E.g. return an ID (string) of the put document if successful but embed the PutResponse in the exception if the operation fails after maximum retries.